logstash-core/src/main/java/org/logstash/ObjectMappers.java

```diff
@@ -107,21 +122,82 @@ public void serializeWithType(final RubyString value, final JsonGenerator jgen,
                 typeSer.typeId(value, RubyString.class, JsonToken.VALUE_STRING);
             typeSer.writeTypePrefix(jgen, typeId);
             final ByteList bytes = value.getByteList();
-            jgen.writeBinary(bytes.getUnsafeBytes(), 0, bytes.length());
+            final byte[] encoding = ENCODING_PREFIXES.computeIfAbsent(
+                value.getEncoding().getClass(), cls -> cls.getName().getBytes(StandardCharsets.UTF_8)
+            );
+            final ByteArrayOutputStream baos =
+                new ByteArrayOutputStream(1 + bytes.length() + Integer.BYTES + encoding.length) {
+                    @Override
+                    public byte[] toByteArray() {
+                        return buf;
+                    }
+                };
+            baos.write(0xff);
+            final int prefixLen = encoding.length;
+            baos.write((byte) prefixLen >> 24);
+            baos.write((byte) prefixLen >> 16);
+            baos.write((byte) prefixLen >> 8);
+            baos.write((byte) prefixLen);
+            baos.write(encoding);
+            baos.write(bytes.getUnsafeBytes(), bytes.begin(), bytes.length());
+            jgen.writeBinary(baos.toByteArray(), 0, baos.size());
             typeSer.writeTypeSuffix(jgen, typeId);
         }
     }
 
     public static final class RubyStringDeserializer extends StdDeserializer<RubyString> {
 
+        private static final Map<ByteBuffer, Encoding> ENCODING_PREFIXES =
+            new ConcurrentHashMap<>();
+
         RubyStringDeserializer() {
             super(RubyString.class);
         }
 
         @Override
         public RubyString deserialize(final JsonParser p, final DeserializationContext ctxt)
             throws IOException {
-            return RubyString.newString(RubyUtil.RUBY, p.getBinaryValue());
+            final byte[] data = p.getBinaryValue();
+            final RubyString result;
+            // If the first unsigned byte in the data is 0xff we have to decode the encoding.
+            if ((data[0] & 0xff) == 0xff) {
```

Comment:
why the bitwise-AND here?

Reply:
@yaauie basically just for readability. I want to compare to compare `date[0]` to `0xff` as unsigned `byte` and this looked a little easier on the eyes than `data[0] == (byte) 0xff`, which just feels wrong.